### DISCUSSION TITLE
field validation

### DISCUSSION THREAD

**Discussion by pavel916** [keycloakify/keycloakify#546] [Q&A] [answered]
Please tell me, is it possible to use functions from react-hook-form to validate fields in the registration component? Will this work?

**Comment by garronej** [comment]
Hello @pavel916,

You can use react-hook-forms, but I advise against it. Keycloak has its own format for validating form fields that is already very versatile. Keycloakify implements client-side field validation based on Keycloak server parameters and does so out of the box. Therefore, there's really no need to use react-hook-forms; you'll only end up hardcoding validation rules that should not be defined in the theme but on the Keycloak server itself.

I suggest reading the documentation here: https://docs.keycloakify.dev/realtime-input-validation

Also, be aware that a significant Keycloakify update is on the horizon, which will dramatically improve form validation based on Keycloak user attributes. It will also validate password policies defined on the server. See more at https://github.com/keycloakify/keycloakify/pull/538

With the Keycloak User Profile, you can define regular expressions to validate fields, format numbers, use multivalued inputs, choose input types (radio buttons, select, textarea, etc.), and display helper text and error messages, among other features.

You might want to wait for this update before starting your project. In any case, I'll publish a migration guide. It won't be dramatically different, just much more validation option will be implemented.

**Reply by pavel916** [reply]
Thank you, sorry for the question, our registration form fields are validated on the Keycloak server only after the user clicks the “register” button. Until this point, for example, loss of focus in an empty field, how to handle it using the useFormValidation hook?

**Reply by garronej** [reply]
What validation in particular ?

**Reply by pavel916** [reply]
for example, if the user has not filled out a required field, he needs to be prompted to do so before he clicks on the 'register' button

**Reply by garronej** [reply, marked as answer]
Well you can see on the deployment of the starter project that the register button is grayed out untill all requiered field are filled up and valid:  

https://starter.keycloakify.dev/  


https://github.com/keycloakify/keycloakify/assets/6702424/e10c9204-8360-418a-8fd2-9b285d027b4c

The requiered field that are not filled up should be in red after first focus lost but this is rather a detail that will be fixed in the next release.

**Comment by pavel916** [comment]
Thanks, and if there is no network, what page should "error.ftl" be shown?

**Reply by pavel916** [reply]
I mean, how should you process a case when there is no network? "...ftl"

**Reply by garronej** [reply]
I mean I'm not sure what you're reffering to by "no network" can you elaborate?

**Comment by pavel916** [comment]
Did the screenshot show how to handle such a scenario? Is it possible to show some component when there is no Internet?
<img width="1435" alt="Снимок экрана 2024-05-07 в 19 46 19" src="https://github.com/keycloakify/keycloakify/assets/105870464/2c0959c7-fdc8-4405-baa1-0b6946e6be84">

**Reply by garronej** [reply]
Well if there is no internet, there is no internet.  
Every navigation on the Keycloak pages requires a back and forth between the Keycloak server and the client side application.  
You might be under the impression that your Keycloak theme is a single page application but it's not. Try and go on the starter deloyment. https://starter.keycloakify.dev/. Navigate between the login and the register page. You'll se a full page refresh. It's Keycloak sending down the HTML it's not client side navigation. Thus handling the no internet scenario is a dead end at this level.    

If you actually want to implement a specific user experience when there's no internet or when the Keycloak server is unreachable you could do so in your react SPA. (Assuming it's what your building).  
It can be relevent if your building a progressive web app that have offline capabilites.   

Checkout https://www.oidc-spa.dev/

You can use the `initializationError` to implement a behavior when the Keycloak server is down or unreachable:  
https://docs.oidc-spa.dev/documentation/error-management

Oidc-spa is from us and it's used in the keycloakify starter.

**Reply by pavel916** [reply]
Thank you very much for your help!
